File: android/app/src/main/java/sol/vortexx/android/spaces/impl/HttpSpacesRepository.kt
package sol.vortexx.android.spaces.impl

import io.ktor.client.call.body
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable
import sol.vortexx.android.db.dao.SpaceDao
import sol.vortexx.android.db.entities.SpaceEntity
import sol.vortexx.android.net.impl.VortexHttpClient
import sol.vortexx.android.spaces.api.SpacesRepository
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpSpacesRepository @Inject constructor(
    private val http: VortexHttpClient,
    private val dao: SpaceDao,
) : SpacesRepository {

    override fun observe(): Flow<List<SpaceEntity>> = dao.observeAll()

    override suspend fun refresh(): Boolean = runCatching {
        val resp = http.client.get("api/spaces/my")
        if (!resp.status.isSuccess()) return@runCatching false
        val body = resp.body<List<SpaceDto>>()
        dao.upsertAll(body.map(SpaceDto::toEntity))
        true
    }.getOrDefault(false)

    override suspend fun create(name: String, isPublic: Boolean): SpaceEntity? = runCatching {
        val resp = http.client.post("api/spaces") {
            contentType(ContentType.Application.Json)
            setBody(CreateReq(name, isPublic))
        }
        if (!resp.status.isSuccess()) return@runCatching null
        val s = resp.body<SpaceDto>().toEntity()
        dao.upsertAll(listOf(s)); s
    }.getOrNull()

    override suspend fun leave(id: Long): Boolean = runCatching {
        val resp = http.client.delete("api/spaces/$id/leave")
        if (resp.status.isSuccess()) { dao.delete(id); true } else false
    }.getOrDefault(false)

    @Serializable private data class CreateReq(val name: String, val is_public: Boolean)
    @Serializable private data class SpaceDto(
        val id: Long, val name: String,
        val creator_id: Long, val avatar_emoji: String = "\uD83C\uDF0C",
        val member_count: Int = 0, val is_public: Boolean = false,
    ) {
        fun toEntity() = SpaceEntity(
            id = id, name = name, ownerId = creator_id,
            avatarEmoji = avatar_emoji, memberCount = member_count, isPublic = is_public,
        )
    }
}

File: ios/Modules/Sources/Spaces/impl/HttpSpacesRepository.swift
import Foundation
import Net
import DB

public final class HttpSpacesRepository: SpacesRepository {
    private let http: HttpClient
    private let dao: SpaceDao

    public init(http: HttpClient, dao: SpaceDao) { self.http = http; self.dao = dao }

    public var spaces: AsyncStream<[Space]> {
        AsyncStream { cont in
            let inner = Task {
                for await rows in dao.observeAll() {
                    cont.yield(rows.map { $0.toDomain })
                }
            }
            cont.onTermination = { @Sendable _ in inner.cancel() }
        }
    }

    public func refresh() async -> Bool {
        do {
            let list = try await http.send(.get("api/spaces/my"), [SpaceDto].self)
            try await dao.upsertAll(list.map(\.toRecord))
            return true
        } catch { return false }
    }

    public func create(name: String, isPublic: Bool) async -> Space? {
        do {
            let req = try HttpRequest.postJson("api/spaces",
                body: CreateReq(name: name, is_public: isPublic))
            let dto = try await http.send(req, SpaceDto.self)
            try await dao.upsertAll([dto.toRecord])
            return dto.toRecord.toDomain
        } catch { return nil }
    }

    public func leave(_ id: Int64) async -> Bool {
        do {
            _ = try await http.send(.delete("api/spaces/\(id)/leave"))
            try await dao.delete(id)
            return true
        } catch { return false }
    }

    private struct CreateReq: Encodable { let name: String; let is_public: Bool }
    private struct SpaceDto: Decodable {
        let id: Int64
        let name: String
        let creator_id: Int64
        let avatar_emoji: String?
        let member_count: Int?
        let is_public: Bool?
        var toRecord: SpaceRecord {
            SpaceRecord(id: id, name: name, ownerId: creator_id,
                        avatarEmoji: avatar_emoji ?? "🌌",
                        memberCount: member_count ?? 0,
                        isPublic: is_public ?? false)
        }
    }
}

private extension SpaceRecord {
    var toDomain: Space {
        Space(id: id, name: name, ownerId: ownerId,
              avatarEmoji: avatarEmoji, memberCount: memberCount, isPublic: isPublic)
    }
}
